### PR TITLE
Filter Update

### PR DESCRIPTION
**Pull request by Ekwav**
Started to implement #100 
Could someone please tell me how you get a datepicker in react?

### CODE REVIEW

File: src/api/ApiHelper.tsx [hunk 1]
```diff
@@ -85,8 +85,7 @@ function initAPI(): API {
             let requestData = {
                 name: itemTagOrName,
                 start: Math.round(fetchStart / 100000) * 100,
-                reforge: itemFilter?.reforge ? itemFilter.reforge.id : undefined,
-                enchantments: itemFilter && itemFilter.enchantment !== undefined && itemFilter.enchantment.level !== undefined && itemFilter.enchantment.id ? [[itemFilter.enchantment.id, itemFilter.enchantment.level]] : undefined
+                filter:itemFilter
```
Comment (matthias-luger) [inline src/api/ApiHelper.tsx:88]: add spaces (similar style across the project)

File: src/api/ApiHelper.tsx [hunk 2]
```diff
@@ -494,14 +493,13 @@ function initAPI(): API {
...
             };
             httpApi.sendLimitedCacheRequest({
                 type: RequestType.RECENT_AUCTIONS,
                 data: requestData,
                 resolve: (data: any) => {
-                    resolve(data.map(a => parseRecentAuction(a)));
+                    resolve(data?.map(a => parseRecentAuction(a)));
```
Comment (matthias-luger) [inline src/api/ApiHelper.tsx:502]: why the ?
shouldt it be save to assume a array? If not check for a empty response and return []

`if(!data){return [])`

File: src/api/ApiHelper.tsx [hunk 3]
```diff
@@ -542,6 +540,21 @@ function initAPI(): API {
...
+    let getFilter = (name:string): Promise<FilterOptions> => {
+        return new Promise((resolve, reject) => {
+            httpApi.sendRequest({
+                type: RequestType.GET_FILTER,
+                data: name,
+                resolve: (data: any) => {
+                    resolve(data);
+                },
+                reject: (error: any) => {
+                    apiErrorHandler(RequestType.GET_FILTER, error, "");
```
Comment (matthias-luger) [inline src/api/ApiHelper.tsx:552]: the third parameter should be the data -> in this case "name"

File: src/api/HttpHelper.tsx [hunk 4]
```diff
@@ -36,8 +36,29 @@ function sendRequest(request: ApiRequest): Promise<void> {
...
+                try {
+                    parsed = response.json();
+                } catch (error) {
+                    return;
+                }
+                return parsed;
+            }).then(parsedResponse => {
+
+                if (!parsedResponse) {
+                    return;
```
Comment (matthias-luger) [inline src/api/HttpHelper.tsx:59]: call request.reject()

File: src/components/FilterElement/FilterElement.tsx [hunk 5]
```diff
@@ -0,0 +1,109 @@
+/* eslint-disable react-hooks/exhaustive-deps */
+/* eslint-disable jsx-a11y/anchor-is-valid */
+import React, { ChangeEvent, useEffect, useRef, useState } from 'react';
+import { Form, Spinner } from 'react-bootstrap';
+import "react-datepicker/dist/react-datepicker.css";
+import './FilterElement.css';
+import api from '../../api/ApiHelper';
+import DatePicker from "react-datepicker";
+
+// has to be redefined because global types from react-app-env are not accessable
```
Comment (matthias-luger) [inline src/components/FilterElement/FilterElement.tsx:10]: are you sure? Why not?

File: src/components/FilterElement/FilterElement.tsx [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+// Boolean if the component is mounted. Set to false in useEffect cleanup function
+let mounted = true;
+
+function FilterElement(props: Props) {
+    let [date, setDate] = useState(new Date())
+    let [value, setValue] = useState<string>();
+
+    useEffect(() => {
+        mounted = true;
+        console.log(props.options)
```
Comment (matthias-luger) [inline src/components/FilterElement/FilterElement.tsx:35]: remove console.log

File: src/components/FilterElement/FilterElement.tsx [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        if (props.options && hasFlag(props.options!.type, FilterTypeEnum.DATE) && props.value)
```
Comment (matthias-luger) [inline src/components/FilterElement/FilterElement.tsx:36]: Typescript should notice your prop.options check. The "!" shouldnt be necessary

Comment (matthias-luger) [inline src/components/FilterElement/FilterElement.tsx:36]: please use {}

File: src/components/FilterElement/FilterElement.tsx [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+            setDate(new Date(parseInt(props.value) * 1000));
+        else
+            setValue(props.value);
+        return () => { mounted = false }
+    }, []);
+
+    let hasFlag = (full?: FilterType, flag?: FilterTypeEnum) => {
+        return full && flag && (full & flag) === flag;
```
Comment (matthias-luger) [inline src/components/FilterElement/FilterElement.tsx:44]: add comment what witchcraft you are doing here

File: src/components/FilterElement/FilterElement.tsx [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                </div>
+            }
+
+        </div >
+    )
+
+    function updateValue(value: string) {
+        let newFilter = {};
+        newFilter[props.options!.name] = value;
+        console.log(newFilter);
```
Comment (matthias-luger) [inline src/components/FilterElement/FilterElement.tsx:102]: remove console.log

File: src/components/FilterElement/FilterElement.tsx [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+                                    {selectOptions}
+                                </Form.Control>
+                    }
+                </div>
+            }
+
+        </div >
+    )
+
+    function updateValue(value: string) {
```
Comment (matthias-luger) [inline src/components/FilterElement/FilterElement.tsx:99]: put the function above the html

File: src/components/ItemFilter/ItemFilter.tsx [hunk 11]
```diff
@@ -47,77 +50,34 @@ function ItemFilter(props: Props) {
         setIsApplied(false);
     })
 
-    let loadFilterItems = () => {
-        api.getEnchantments().then(enchantments => {
-            if (!mounted) {
+    let loadFilterOptions = (filterName: string) => {
+        if (filterOptions.some(el => el.name == filterName))
```
Comment (matthias-luger) [inline src/components/ItemFilter/ItemFilter.tsx:54]: use {}

File: src/components/ItemFilter/ItemFilter.tsx [hunk 12]
```diff
@@ -47,77 +50,34 @@ function ItemFilter(props: Props) {
...
+            return;
+        api.getFilter(filterName).then(options => {
+            console.log(props);
```
Comment (matthias-luger) [inline src/components/ItemFilter/ItemFilter.tsx:57]: remove console.log

File: src/components/ItemFilter/ItemFilter.tsx [hunk 13]
```diff
@@ -47,77 +50,34 @@ function ItemFilter(props: Props) {
...
+            console.log(options)
```
Comment (matthias-luger) [inline src/components/ItemFilter/ItemFilter.tsx:58]: remove console.log

File: src/components/ItemFilter/ItemFilter.tsx [hunk 14]
```diff
@@ -47,77 +50,34 @@ function ItemFilter(props: Props) {
...
+            if (filterOptions.some(el => el.name == filterName))
```
Comment (matthias-luger) [inline src/components/ItemFilter/ItemFilter.tsx:59]: use {}

File: src/components/ItemFilter/ItemFilter.tsx [hunk 15]
```diff
@@ -47,77 +50,34 @@ function ItemFilter(props: Props) {
...
-                        reforge: {
-                            id: reforges[0].id,
-                            name: reforges[0].name
-                        }
-                    })
-                }
-            })
+            filterOptions.push(options);
+            setFilterOptions(filterOptions);
+            //setSelectedFilters([...selectedFilters]);
```
Comment (matthias-luger) [inline src/components/ItemFilter/ItemFilter.tsx:63]: remove comment

File: src/components/ItemFilter/ItemFilter.tsx [hunk 16]
```diff
@@ -47,77 +50,34 @@ function ItemFilter(props: Props) {
...
 
-    let onLevelChange = (newLevel: ChangeEvent) => {
-        let newFilter: ItemFilter = {
-            enchantment: {
-                id: itemFilter?.enchantment?.id || 0,
-                level: parseInt((newLevel.target as HTMLInputElement).value)
-            },
-            reforge: itemFilter?.reforge
-        };
+        //itemFilter![filterName] = "";
```
Comment (matthias-luger) [inline src/components/ItemFilter/ItemFilter.tsx:73]: remove comment

File: src/components/ItemFilter/ItemFilter.tsx [hunk 17]
```diff
@@ -153,25 +111,38 @@ function ItemFilter(props: Props) {
...
+    let filterSelectList = props?.filters ? props?.filters.filter(f => !selectedFilters.includes(f)).map(filter => {
         return (
-            <option data-id={reforge.id} key={reforge.id} value={reforge.name}>{reforge.name}</option>
+            <option data-id={filter} key={filter} value={filter}>{filter}</option>
         )
-    })
+    }) : ""
+
+    let onFilterChange = (filter?: ItemFilter) => {
+        console.log(filter);
```
Comment (matthias-luger) [inline src/components/ItemFilter/ItemFilter.tsx:123]: remove console.log

File: src/components/ItemFilter/ItemFilter.tsx [hunk 18]
```diff
@@ -153,25 +111,38 @@ function ItemFilter(props: Props) {
...
+        console.log(Object.keys(filter as object));
```
Comment (matthias-luger) [inline src/components/ItemFilter/ItemFilter.tsx:124]: remove console.log

File: src/components/ItemFilter/ItemFilter.tsx [hunk 19]
```diff
@@ -153,25 +111,38 @@ function ItemFilter(props: Props) {
...
+
+        var keys = Object.keys(filter as object);
+        if (keys.length > 0) {
+            var key = keys[0];
+            itemFilter![key] = filter![key];
+            console.log("udpated");
```
Comment (matthias-luger) [inline src/components/ItemFilter/ItemFilter.tsx:130]: remove console.log

File: src/components/ItemFilter/ItemFilter.tsx [hunk 20]
```diff
@@ -153,25 +111,38 @@ function ItemFilter(props: Props) {
...
+    }) : ""
+
+    let onFilterChange = (filter?: ItemFilter) => {
+        console.log(filter);
+        console.log(Object.keys(filter as object));
+
+        var keys = Object.keys(filter as object);
+        if (keys.length > 0) {
+            var key = keys[0];
+            itemFilter![key] = filter![key];
```
Comment (matthias-luger) [inline src/components/ItemFilter/ItemFilter.tsx:129]: can you be sure itemFilter and filter exist?
check with a if

File: src/components/PriceGraph/PriceGraph.tsx [hunk 21]
```diff
@@ -88,6 +89,8 @@ function PriceGraph(props: Props) {
 
             priceChart.update();
             setIsFilterable(result.filterable);
+            console.log(result.filters)
```
Comment (matthias-luger) [inline src/components/PriceGraph/PriceGraph.tsx:92]: remove console.log

File: src/utils/Parser/APIResponseParser.tsx [hunk 22]
```diff
@@ -81,9 +81,10 @@ export function parseItemPrice(priceData: any): ItemPrice {
 export function parseItemPriceData(priceData: any): ItemPriceData {
     return {
         filterable: priceData.filterable,
-        prices: priceData.prices.map(price => {
+        prices: priceData.prices?.map(price => {
```
Comment (matthias-luger) [inline src/utils/Parser/APIResponseParser.tsx:84]: whe the ?

**Review by matthias-luger — CHANGES_REQUESTED** [review]
see comments

File: src/components/FilterElement/FilterElement.tsx (hunk 5 above)
Comment (Ekwav) [reply src/components/FilterElement/FilterElement.tsx:10]: typescript doesn't deliver the global types definition file.

File: src/components/ItemFilter/ItemFilter.tsx (hunk 20 above)
Comment (Ekwav) [reply src/components/ItemFilter/ItemFilter.tsx:146]: it does exist

File: src/utils/Parser/APIResponseParser.tsx (hunk 22 above)
Comment (Ekwav) [reply src/utils/Parser/APIResponseParser.tsx:84]: because prices could be null

File: src/api/HttpHelper.tsx [hunk 23]
```diff
@@ -36,10 +36,26 @@ function sendRequest(request: ApiRequest): Promise<void> {
 
         requests.push(request);
         return fetch(url, { headers })
-            .then(response => response.json())
-            .then(parsedResponse => {
-                if (parsedResponse.type === "error") {
-                    toast.error(JSON.parse(parsedResponse.data).data);
+            .then(response => {
+                if (response.ok && response.body === null) {
```
Comment (matthias-luger) [inline src/api/HttpHelper.tsx:40]: extract all this if-conditions in a function checkForValidResponse()

Comment (matthias-luger) [inline src/api/HttpHelper.tsx:40]: extract all this if-conditions in a function checkForValidResponse()

File: src/components/FilterElement/FilterElement.css [hunk 24]
```diff
@@ -0,0 +1,19 @@
+.enchantment-filter {
```
Comment (matthias-luger) [inline src/components/FilterElement/FilterElement.css:1]: remove unused css

File: src/components/FilterElement/FilterElement.tsx [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+/* eslint-disable jsx-a11y/anchor-is-valid */
+import React, { ChangeEvent, useEffect, useRef, useState } from 'react';
+import { Form, Spinner } from 'react-bootstrap';
+import "react-datepicker/dist/react-datepicker.css";
+import './FilterElement.css';
+import api from '../../api/ApiHelper';
+import DatePicker from "react-datepicker";
+
+// has to be redefined because global types from react-app-env are not accessable
+export enum FilterTypeEnum {
```
Comment (matthias-luger) [inline src/components/FilterElement/FilterElement.tsx:11]: move the enum declaration into another file (z.B. FilterTypes.tsx)

File: src/react-app-env.d.ts [hunk 26]
```diff
@@ -87,12 +87,27 @@ interface PlayerDetails {
   bids: ItemBid[];
   auctions: Auction[];
 }
+enum FilterType
```
Comment (matthias-luger) [inline src/react-app-env.d.ts:90]: remove this (it should be in another file after the previous task)

**Review by matthias-luger — CHANGES_REQUESTED** [review]

File: src/react-app-env.d.ts (hunk 26 above)
Comment (Ekwav) [reply src/react-app-env.d.ts:90]: leads to compiler errors.

Comment (Ekwav) [reply src/react-app-env.d.ts:90]: can't be removed the type below depends on it. 
And react-app.env.d.ts can't import other files

File: src/api/HttpHelper.tsx [hunk 27]
```diff
@@ -36,10 +36,26 @@ function sendRequest(request: ApiRequest): Promise<void> {
 
         requests.push(request);
         return fetch(url, { headers })
-            .then(response => response.json())
-            .then(parsedResponse => {
-                if (parsedResponse.type === "error") {
-                    toast.error(JSON.parse(parsedResponse.data).data);
+            .then(response => {
```
Comment (matthias-luger) [inline src/api/HttpHelper.tsx:39]: put all if checks into the checkForValidResponse

**Review by matthias-luger — CHANGES_REQUESTED** [review]

**Comment by matthias-luger** [comment]
the current state instantly craches
-> ApiHelper:19 check if error is undefined and if it has error.Message
-> getFilter seems to not return a error-Object for me

**Comment by Ekwav** [comment]
> the current state instantly craches
> -> ApiHelper:19 check if error is undefined and if it has error.Message
> -> getFilter seems to not return a error-Object for me

**Comment by Ekwav** [comment]
> the current state instantly craches
> -> ApiHelper:19 check if error is undefined and if it has error.Message
> -> getFilter seems to not return a error-Object for me

@matthias-luger  it is unclear to me what the last line means

**Comment by matthias-luger** [comment]
image
Please improve the styling. 

- Make the "remove"-Button a real button.
- Style the date-Field similar to the select field (label on top, width, ...)

**Review by matthias-luger — CHANGES_REQUESTED** [review]
image

- Make the "remove"-Button a real button.
- Style the date-Field similar to the select field (label on top, width, ...)
- improve general styling
- Add cypress tests (optional)
- change wording to "Add Filter" not "Click to select filter"
- changing the timespan seems to break the filter
- Uncaught (in promise) SyntaxError: Unexpected token u in JSON at position 0 at JSON.parse (<anonymous>)at HttpHelper.tsx:54
- url didnt save the petlevel

**Comment by Ekwav** [comment]
@matthias-luger your fiexes introduced bugs 
* The graph is created multiple times wich leads to a flickering when hovered.
* Filters from the url are not displayed (Add `Reforge` filter, apply, reload, filter isn't displayed and marked as not applied)
* DateFilter has no default value and site crashes when a value is selected
There is an error message in the console
```
index.js:1 Warning: The final argument passed to useEffect changed size between renders. The order and size of this array must remain constant.

Previous: []
Incoming: [Reforge, Enchantment, EnchantLvl, UId, EndBefore, EndAfter]
    in ItemFilter (at PriceGraph.tsx:142)
    in div (at PriceGraph.tsx:141)
```

**Review by matthias-luger — APPROVED** [review]

**Comment by matthias-luger** [comment]
@Ekwav please test it again it should work now

**Comment by Ekwav** [comment]
Thanks @matthias-luger 
A few bugs are still in there.
* When there no recent auction for a filter the old auctions are displayed
* The `Filter is currently NOT applied` is displayed when the page is opened via a url containing a filter. It replaces the `Applied` badge after a few milliseconds 
* The filter date range should be set to `day` when the filter changes
* Date filters now have a default value but that value isn't used. The date filter value only gets added to the filter object when it changes

**Pull request merged**